internal/config/config.go

```diff
@@ -319,9 +325,35 @@ func peerFromCR(p metallbv1beta1.BGPPeer) (*Peer, error) {
 		}
 	}
 
+	if p.Spec.Password != "" && p.Spec.SecretRef != nil {
+		return nil, fmt.Errorf("can not have both password and secret ref set in the peer config")
+	}
+
 	var password string
 	if p.Spec.Password != "" {
 		password = p.Spec.Password
+	} else if p.Spec.SecretRef != nil {
+		secret := &corev1.Secret{}
+		err := cli.Get(ctx, types.NamespacedName{Namespace: p.Namespace, Name: p.Spec.SecretRef.Name}, secret)
+		if err != nil {
+			return nil, err
+		}
+		if secret.Type == corev1.SecretTypeOpaque || secret.Type == corev1.SecretTypeBasicAuth {
+			var (
+				srcPass []byte
+				ok      bool
+			)
+			if srcPass, ok = secret.Data["password"]; !ok {
+				return nil, fmt.Errorf("password not specified in the secret %q", p.Spec.SecretRef)
+			}
+			dstPass, err := base64.StdEncoding.DecodeString(string(srcPass))
+			if err != nil {
+				return nil, err
+			}
+			password = string(dstPass)
+		} else {
+			return nil, fmt.Errorf("invalid secret specified in the secret %q", p.Spec.SecretRef)
```

Comment:
can we give more info in the err msg? maybe mentioning the type mismatch.

Reply:
sure, done.